File: echoback.cc
// Echoback (C++): 最小構成のローカル・エコーバック + スイッチ式エコーサプレッサ
// Usage:
//   ./echoback [--passthrough]
// 前提:
//   - 16 kHz モノラル固定, 16-bit I/O (PortAudio デフォルトデバイス)
//   - エコーサプレッサは 10ms（160サンプル）単位で処理
//   - 参照信号は直前にスピーカへ出したブロック（ローカル・ループバック相当）

#include <portaudio.h>

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <deque>
#include <string>
#include <vector>

#include "suppressor.h"

constexpr int kSampleRateHz = 16000;
constexpr int kBlockLen = 160;  // 10 ms block at 16 kHz

 

struct State {
  // 1ch 固定（PortAudioデバイス設定も1ch）
  // ジッタバッファは固定遅延（最小構成: 内部固定値）

  // FIFOs in device domain (dev_sr)
  std::deque<int16_t> rec_dev;     // mic captured samples
  std::deque<int16_t> out_dev;     // to speaker

  // 処理フレーム = デバイスサンプルレート領域（10ms = 160サンプル）

  // Jitter buffer (device domain). Local echo path accumulator, mixes into speaker
  std::deque<int16_t> jitter;      // accumulate processed to emulate loopback latency

  // Optional startup delay for capture stream (educational jitter buffer)
  std::deque<int16_t> delay_line;  // raw capture samples waiting for release
  size_t delay_target_samples = 0;  // optional capture-side delay in samples (default off)

  // Optional extra delay applied to far-end reference before suppression
  std::deque<int16_t> far_delay_line;  // buffered far samples for loopback delay simulation
  size_t loopback_delay_target_samples = 0;  // additional far-end delay in samples

  // --passthrough: AEC を行わず素通し再生
  bool passthrough = false;
  EchoSuppressor suppressor;
  size_t block_counter = 0;

  // 遅延推定の観察用統計
  std::deque<int> lag_history;
  long long lag_sum = 0;
  size_t lag_history_limit = 10;
  bool lag_stats_ready = false;
  double lag_avg_latest = 0.0;
  int lag_min_window = 0;
  int lag_max_window = 0;
  int lag_last = -1;

  // 単一インスタンス化API。個別インスタンスは不要。
};

size_t pop_samples(std::deque<int16_t>& q, int16_t* dst, size_t n){
  size_t m = q.size()<n ? q.size() : n;
  for (size_t i=0;i<m;i++){ dst[i]=q.front(); q.pop_front(); }
  return m;
}

void push_block(std::deque<int16_t>& q, const int16_t* src, size_t n){
  for (size_t i=0;i<n;i++) q.push_back(src[i]);
}

inline void enqueue_capture_sample(State& s, int16_t sample) {
  s.delay_line.push_back(sample);
  if (s.delay_line.size() <= s.delay_target_samples) {
    s.rec_dev.push_back(0);
  } else {
    s.rec_dev.push_back(s.delay_line.front());
    s.delay_line.pop_front();
  }
}

inline void apply_loopback_delay(State& s, int16_t* block) {
  const size_t target = s.loopback_delay_target_samples; // 回り込み音を意図的に遅らせる
  if (target == 0) return;

  for (int i = 0; i < kBlockLen; ++i) {
    s.far_delay_line.push_back(block[i]);
    int16_t delayed = 0;
    if (s.far_delay_line.size() > target) {
      delayed = s.far_delay_line.front();
      s.far_delay_line.pop_front();
    }
    block[i] = delayed;
  }
}

void process_available_blocks(State& s){
  // Run as many 10 ms (160-sample) blocks as possible
  while (s.rec_dev.size() >= static_cast<size_t>(kBlockLen)) {
    std::vector<int16_t> near_blk(kBlockLen), far_blk(kBlockLen), out_blk(kBlockLen);
    // 1) pop near block
    pop_samples(s.rec_dev, near_blk.data(), kBlockLen);
    // 2) pop far block（スピーカへ送る信号 = 参照）
    if (s.jitter.size() >= static_cast<size_t>(kBlockLen)) {
      pop_samples(s.jitter, far_blk.data(), kBlockLen);
    } else {
      std::fill(far_blk.begin(), far_blk.end(), 0);
    }
    std::vector<int16_t> speaker_blk = far_blk;  // スピーカへ送る信号は個別に遅延する

    apply_loopback_delay(s, speaker_blk.data());

    float gate_gain = 1.0f;
    int estimated_lag = 0;
    if (s.passthrough) {
      std::memcpy(out_blk.data(), near_blk.data(), kBlockLen * sizeof(int16_t));
    } else {
      static constexpr float kInvScale = 1.0f / 32768.0f;
      static constexpr float kScale = 32767.0f;
      std::vector<float> far_block_f(kBlockLen), near_block_f(kBlockLen), out_block_f(kBlockLen);
      for (int i = 0; i < kBlockLen; ++i) {
        far_block_f[i] = static_cast<float>(far_blk[i]) * kInvScale;
        near_block_f[i] = static_cast<float>(near_blk[i]) * kInvScale;
      }
      s.suppressor.process_block(far_block_f.data(),
                                 near_block_f.data(),
                                 out_block_f.data(),
                                 &gate_gain,
                                 &estimated_lag);
      for (int i = 0; i < kBlockLen; ++i) {
        float sample = std::max(-1.0f, std::min(1.0f, out_block_f[i]));
        out_blk[i] = static_cast<int16_t>(std::lrintf(sample * kScale));
      }
    }

    if (estimated_lag >= 0) {
      s.lag_history.push_back(estimated_lag);
      s.lag_sum += estimated_lag;
      if (s.lag_history.size() > s.lag_history_limit) {
        s.lag_sum -= s.lag_history.front();
        s.lag_history.pop_front();
      }
      const size_t window_size = s.lag_history.size();
      if (window_size > 0) {
        int min_val = s.lag_history.front();
        int max_val = s.lag_history.front();
        for (int lag_sample : s.lag_history) {
          if (lag_sample < min_val) {
            min_val = lag_sample;
          }
          if (lag_sample > max_val) {
            max_val = lag_sample;
          }
        }
        s.lag_avg_latest = static_cast<double>(s.lag_sum) / static_cast<double>(window_size);
        s.lag_min_window = min_val;
        s.lag_max_window = max_val;
        s.lag_last = estimated_lag;
        s.lag_stats_ready = true;
      }
    }
    const size_t lag_window_size = s.lag_history.size();

    if (!s.passthrough) {
      float mute_ratio = 1.0f - gate_gain;
      if (mute_ratio < 0.0f) mute_ratio = 0.0f;
      const char* gain_meter = GainMeterString(gate_gain);
      if (s.lag_stats_ready) {
        char lag_current_buf[32];
        const char* lag_current_str = "--";
        if (estimated_lag >= 0) {
          std::snprintf(lag_current_buf, sizeof(lag_current_buf), "%d", estimated_lag);
          lag_current_str = lag_current_buf;
        }
        const size_t avg_window = lag_window_size > 0 ? lag_window_size : s.lag_history_limit;
        std::fprintf(stderr,
                     "[block %zu] mute=%.1f%% (gain=%.3f %s, lag=%s samples; avg%zu=%.1f, min=%d, max=%d, last=%d)\n",
                     s.block_counter,
                     mute_ratio * 100.0f,
                     gate_gain,
                     gain_meter,
                     lag_current_str,
                     avg_window,
                     s.lag_avg_latest,
                     s.lag_min_window,
                     s.lag_max_window,
                     s.lag_last);
      } else {
        std::fprintf(stderr, "[block %zu] mute=%.1f%% (gain=%.3f %s, lag=--)\n",
                     s.block_counter,
                     mute_ratio * 100.0f,
                     gate_gain,
                     gain_meter);
      }
    }
    ++s.block_counter;

    // ローカル・ループバック: 処理後の出力を蓄積して、次々回以降の far にする
    push_block(s.jitter, out_blk.data(), kBlockLen);

    // 今回のスピーカ出力はループバック遅延を適用した信号
    push_block(s.out_dev, speaker_blk.data(), kBlockLen);
  }
}

int pa_callback(const void* inputBuffer,
                void* outputBuffer,
                unsigned long blockSize,
                const PaStreamCallbackTimeInfo* /*timeInfo*/,
                PaStreamCallbackFlags /*statusFlags*/,
                void* userData){
  auto* st = reinterpret_cast<State*>(userData);
  const int16_t* in = reinterpret_cast<const int16_t*>(inputBuffer);
  int16_t* out = reinterpret_cast<int16_t*>(outputBuffer);
  const unsigned long n = blockSize; // mono (framesPerBuffer)
  // Single-threaded実行のため終了フラグは不要

  // 1) enqueue capture
  for (unsigned long i = 0; i < n; ++i) {
    int16_t sample = in ? in[i] : 0;
    enqueue_capture_sample(*st, sample);
  }

  // 2) run AEC3 on as many blocks as ready
  process_available_blocks(*st);

  // 3) emit output
  for (unsigned long i=0;i<n;i++){
    if (!st->out_dev.empty()){ out[i]=st->out_dev.front(); st->out_dev.pop_front(); }
    else { out[i]=0; }
  }
  return paContinue;
}

int main(int argc, char** argv){
  State s;
  // 16k/10msブロック固定

  // 引数パース
  for (int i = 1; i < argc; ++i) {
    std::string arg(argv[i] ? argv[i] : "");
    if (arg == "--passthrough" || arg == "-p") {
      s.passthrough = true;
    } else if (arg.rfind("--input-delay-ms=", 0) == 0) {
      std::string value = arg.substr(strlen("--input-delay-ms="));
      long long delay_ms = std::stoll(value);
      size_t raw_samples = static_cast<size_t>(
          (delay_ms * static_cast<long long>(kSampleRateHz) + 999) / 1000);
      const size_t block = static_cast<size_t>(kBlockLen);
      size_t delay_blocks = (raw_samples + block / 2) / block;
      s.delay_target_samples = delay_blocks * block;
    } else if (arg.rfind("--loopback-delay-ms=", 0) == 0) {
      std::string value = arg.substr(strlen("--loopback-delay-ms="));
      long long delay_ms = std::stoll(value);
      if (delay_ms < 0) delay_ms = 0;
      size_t raw_samples = static_cast<size_t>(
          (delay_ms * static_cast<long long>(kSampleRateHz) + 999) / 1000);
      s.loopback_delay_target_samples = raw_samples;
      s.far_delay_line.clear();
    } else if (arg == "--input-delay-ms" && i + 1 < argc) {
      std::string value(argv[++i] ? argv[i] : "0");
      long long delay_ms = std::stoll(value);
      size_t raw_samples = static_cast<size_t>(
          (delay_ms * static_cast<long long>(kSampleRateHz) + 999) / 1000);
      const size_t block = static_cast<size_t>(kBlockLen);
      size_t delay_blocks = (raw_samples + block / 2) / block;
      s.delay_target_samples = delay_blocks * block;
    } else if (arg == "--loopback-delay-ms" && i + 1 < argc) {
      std::string value(argv[++i] ? argv[i] : "0");
      long long delay_ms = std::stoll(value);
      if (delay_ms < 0) delay_ms = 0;
      size_t raw_samples = static_cast<size_t>(
          (delay_ms * static_cast<long long>(kSampleRateHz) + 999) / 1000);
      s.loopback_delay_target_samples = raw_samples;
      s.far_delay_line.clear();
    } else if (arg == "--help" || arg == "-h") {
      std::fprintf(stderr,
                   "Usage: %s [--passthrough] [--input-delay-ms <ms>] [--loopback-delay-ms <ms>]\n",
                   argv[0]);
      return 0;
    } else {
      std::fprintf(stderr, "Unknown option: %s\n", arg.c_str());
      return 1;
    }
  }
  const char* mode = s.passthrough ? "passthrough" : "suppressor";
  std::fprintf(stderr, "echoback (16k mono): mode=%s\n", mode);
  if (!s.passthrough) {
    std::fprintf(stderr,
                 "  config: atten=%.1f dB, rho=%.2f, ratio=%.2f, hang=%d, attack=%.3f, release=%.3f\n",
                 -80.0f,
                 EchoSuppressor::kRhoThresh,
                 EchoSuppressor::kPowerRatioAlpha,
                 EchoSuppressor::kHangoverBlocks,
                 EchoSuppressor::kAttack,
                 EchoSuppressor::kRelease);
  }
  // 固定設定のため追加初期化不要

  {
    const double input_delay_ms = static_cast<double>(s.delay_target_samples) * 1000.0 /
                                  static_cast<double>(kSampleRateHz);
    const double input_delay_blocks = static_cast<double>(s.delay_target_samples) /
                                      static_cast<double>(kBlockLen);
    std::fprintf(stderr,
                 "input-delay-ms(final): %.1f ms (%zu samples, %.1f blocks)\n",
                 input_delay_ms,
                 s.delay_target_samples,
                 input_delay_blocks);
  }

  {
    const double loopback_delay_ms = static_cast<double>(s.loopback_delay_target_samples) * 1000.0 /
                                     static_cast<double>(kSampleRateHz);
    std::fprintf(stderr,
                 "loopback-delay-ms(final): %.1f ms (%zu samples)\n",
                 loopback_delay_ms,
                 s.loopback_delay_target_samples);
  }

  PaError err = Pa_Initialize();
  if (err!=paNoError){ std::fprintf(stderr, "Pa_Initialize error %s\n", Pa_GetErrorText(err)); return 1; }

  PaStream* stream=nullptr; PaStreamParameters inP{}, outP{};
  inP.device = Pa_GetDefaultInputDevice();
  outP.device = Pa_GetDefaultOutputDevice();
  if (inP.device<0 || outP.device<0){ std::fprintf(stderr, "No default device.\n"); Pa_Terminate(); return 1; }
  inP.channelCount = 1; inP.sampleFormat = paInt16;
  inP.suggestedLatency = Pa_GetDeviceInfo(inP.device)->defaultLowInputLatency;
  outP.channelCount = 1; outP.sampleFormat = paInt16;
  outP.suggestedLatency = Pa_GetDeviceInfo(outP.device)->defaultLowOutputLatency;

  err = Pa_OpenStream(&stream, &inP, &outP, kSampleRateHz, kBlockLen, paClipOff, pa_callback, &s);
  if (err!=paNoError){ std::fprintf(stderr, "Pa_OpenStream error %s\n", Pa_GetErrorText(err)); Pa_Terminate(); return 1; }
  err = Pa_StartStream(stream);
  if (err!=paNoError){ std::fprintf(stderr, "Pa_StartStream error %s\n", Pa_GetErrorText(err)); Pa_CloseStream(stream); Pa_Terminate(); return 1; }

  std::fprintf(stderr, "Running... Ctrl-C to stop.\n");
  while (Pa_IsStreamActive(stream)==1) {
    Pa_Sleep(100);
  }
  Pa_StopStream(stream); Pa_CloseStream(stream);
  Pa_Terminate();
  std::fprintf(stderr, "stopped.\n");
  return 0;
}

File: echoback.js
// Node echoback: PortAudio(PAmac.node) + minES suppressor (16kHz mono, 160-sample blocks)
// Usage: node echoback.js [--passthrough] [--input-delay-ms <ms>] [--loopback-delay-ms <ms>] [latency_ms]

const assert = require('assert');
const { JsSuppressor, gainMeterString, SAMPLE_RATE, BLOCK_SAMPLES } = require('./suppressor');

let passthrough = false;
let jitterLatencyMs = 200;
let inputDelayMs = 0;
let loopbackDelayMs = 150;

function parseMs(value, flag) {
  const n = Number(value);
  if (!Number.isFinite(n) || n < 0) {
    console.error(`${flag} expects non-negative number (ms). got=${value}`);
    process.exit(1);
  }
  return Math.min(100000, n);
}

const args = process.argv.slice(2);
for (let i = 0; i < args.length; i++) {
  const arg = args[i];
  if (arg === '--passthrough' || arg === '-p') {
    passthrough = true;
  } else if (arg === '--help' || arg === '-h') {
    console.error('Usage: node echoback.js [--passthrough] [--input-delay-ms <ms>] [--loopback-delay-ms <ms>] [latency_ms]');
    process.exit(0);
  } else if (arg.startsWith('--input-delay-ms=')) {
    const value = arg.split('=', 2)[1];
    inputDelayMs = parseMs(value, '--input-delay-ms');
  } else if (arg === '--input-delay-ms') {
    if (i + 1 >= args.length) {
      console.error('--input-delay-ms requires a value');
      process.exit(1);
    }
    inputDelayMs = parseMs(args[++i], '--input-delay-ms');
  } else if (arg.startsWith('--loopback-delay-ms=')) {
    const value = arg.split('=', 2)[1];
    loopbackDelayMs = parseMs(value, '--loopback-delay-ms');
  } else if (arg === '--loopback-delay-ms') {
    if (i + 1 >= args.length) {
      console.error('--loopback-delay-ms requires a value');
      process.exit(1);
    }
    loopbackDelayMs = parseMs(args[++i], '--loopback-delay-ms');
  } else if (/^\d+$/.test(arg)) {
    jitterLatencyMs = Math.max(0, Math.min(10000, parseInt(arg, 10)));
  } else {
    console.error('Unknown arg:', arg);
    process.exit(1);
  }
}

const kSr = SAMPLE_RATE;
const kBlock = BLOCK_SAMPLES;

function roundInputDelaySamples(ms) {
  const rawSamples = Math.max(0, Math.round((ms * kSr) / 1000));
  const delayBlocks = Math.floor((rawSamples + Math.floor(kBlock / 2)) / kBlock);
  return delayBlocks * kBlock;
}

function roundDelaySamples(ms) {
  return Math.max(0, Math.floor((ms * kSr + 999) / 1000));
}

const inputDelaySamples = roundInputDelaySamples(inputDelayMs);
const loopbackDelaySamples = roundDelaySamples(loopbackDelayMs);
const jitterSamplesTarget = Math.max(0, Math.floor((kSr * jitterLatencyMs) / 1000));

let PortAudio = null;
if (process.platform === 'darwin') {
  PortAudio = require('./PAmac.node');
} else {
  console.error('Only macOS (PAmac.node) is supported here.');
  process.exit(1);
}

PortAudio.initSampleBuffers(kSr, kSr, kBlock);
PortAudio.startMic();
PortAudio.startSpeaker();

const suppressor = new JsSuppressor();
assert.strictEqual(suppressor.blockSamples, kBlock, 'Unexpected suppressor block size');

console.error(`echoback (16k mono): mode=${passthrough ? 'passthrough' : 'suppressor'}`);
if (!passthrough) {
  console.error('  config: atten=-80.0 dB, rho=0.60, ratio=1.30, hang=20, attack=0.100, release=0.010');
}
const inputDelayMsFinal = inputDelaySamples * 1000 / kSr;
const inputDelayBlocks = kBlock > 0 ? inputDelaySamples / kBlock : 0;
console.error(`input-delay-ms(final): ${inputDelayMsFinal.toFixed(1)} ms (${inputDelaySamples} samples, ${inputDelayBlocks.toFixed(1)} blocks)`);
const loopbackDelayMsFinal = loopbackDelaySamples * 1000 / kSr;
console.error(`loopback-delay-ms(final): ${loopbackDelayMsFinal.toFixed(1)} ms (${loopbackDelaySamples} samples)`);
console.error(`latency buffer target: ${jitterLatencyMs.toFixed(1)} ms (${jitterSamplesTarget} samples)`);
console.error('Running... Ctrl-C to stop.');

function pushArray(dstArr, src) {
  for (let i = 0; i < src.length; i++) dstArr.push(src[i] | 0);
}

function popBlockI16(arr) {
  if (arr.length < kBlock) return null;
  const out = new Int16Array(kBlock);
  for (let i = 0; i < kBlock; i++) out[i] = arr.shift();
  return out;
}

function toInt16Array(x) {
  if (!x) return new Int16Array();
  if (ArrayBuffer.isView(x)) {
    if (x instanceof Int16Array) return x;
    return new Int16Array(x.buffer, x.byteOffset, Math.floor(x.byteLength / 2));
  }
  if (Buffer.isBuffer(x)) {
    return new Int16Array(x.buffer, x.byteOffset, Math.floor(x.byteLength / 2));
  }
  const out = new Int16Array(x.length);
  for (let i = 0; i < x.length; i++) out[i] = x[i] | 0;
  return out;
}

const recQ = [];
const refQ = [];
const jitterQ = [];
const captureDelayLine = [];
let captureDelayStart = 0;
const loopbackDelayLine = [];
let loopbackDelayStart = 0;
let needJitter = jitterSamplesTarget > 0;
let blockCounter = 0;
const lagHistory = [];
let lagSum = 0;
const lagHistoryLimit = 10;
let lagStatsReady = false;
let lagAvgLatest = 0.0;
let lagMinWindow = 0;
let lagMaxWindow = 0;
let lagLast = -1;

function trimCaptureQueue() {
  if (captureDelayStart > 4096 && captureDelayStart > captureDelayLine.length / 2) {
    captureDelayLine.splice(0, captureDelayStart);
    captureDelayStart = 0;
  }
}

function trimLoopbackQueue() {
  if (loopbackDelayStart > 4096 && loopbackDelayStart > loopbackDelayLine.length / 2) {
    loopbackDelayLine.splice(0, loopbackDelayStart);
    loopbackDelayStart = 0;
  }
}

function enqueueCaptureSamples(samples) {
  if (inputDelaySamples === 0) {
    pushArray(recQ, samples);
    return;
  }
  for (let i = 0; i < samples.length; i++) {
    captureDelayLine.push(samples[i] | 0);
    if (captureDelayLine.length - captureDelayStart > inputDelaySamples) {
      recQ.push(captureDelayLine[captureDelayStart++]);
      trimCaptureQueue();
    } else {
      recQ.push(0);
    }
  }
}

function applyLoopbackDelayBlock(block) {
  if (loopbackDelaySamples <= 0) {
    return block;
  }
  const delayed = new Int16Array(block.length);
  for (let i = 0; i < block.length; i++) {
    loopbackDelayLine.push(block[i] | 0);
    if (loopbackDelayLine.length - loopbackDelayStart > loopbackDelaySamples) {
      delayed[i] = loopbackDelayLine[loopbackDelayStart++];
      trimLoopbackQueue();
    } else {
      delayed[i] = 0;
    }
  }
  return delayed;
}

function updateLagStats(lagSamples) {
  if (lagSamples < 0) {
    return;
  }
  lagHistory.push(lagSamples);
  lagSum += lagSamples;
  if (lagHistory.length > lagHistoryLimit) {
    lagSum -= lagHistory.shift();
  }
  if (lagHistory.length > 0) {
    let minVal = lagHistory[0];
    let maxVal = lagHistory[0];
    for (let i = 1; i < lagHistory.length; i++) {
      const v = lagHistory[i];
      if (v < minVal) minVal = v;
      if (v > maxVal) maxVal = v;
    }
    lagAvgLatest = lagSum / lagHistory.length;
    lagMinWindow = minVal;
    lagMaxWindow = maxVal;
    lagLast = lagSamples;
    lagStatsReady = true;
  }
}

function processBlocks() {
  while (recQ.length >= kBlock) {
    const rec = popBlockI16(recQ);
    const ref = refQ.length >= kBlock ? popBlockI16(refQ) : new Int16Array(kBlock);

    let out = rec;
    let gain = 1.0;
    let lagSamples = -1;

    if (!passthrough) {
      const result = suppressor.processInt16(ref, rec);
      out = result.out;
      gain = result.gain;
      lagSamples = result.lag;
      updateLagStats(lagSamples);
      const muteRatio = Math.max(0, Math.min(1, 1 - gain));
      const gainMeter = gainMeterString(gain);
      const currentBlock = blockCounter;
      const lagWindowSize = lagHistory.length;
      if (lagStatsReady) {
        const lagCurrentStr = lagSamples >= 0 ? String(lagSamples) : '--';
        const avgWindow = lagWindowSize > 0 ? lagWindowSize : lagHistoryLimit;
        console.error(`[block ${currentBlock}] mute=${(muteRatio * 100).toFixed(1)}% (gain=${gain.toFixed(3)} ${gainMeter}, lag=${lagCurrentStr} samples; avg${avgWindow}=${lagAvgLatest.toFixed(1)}, min=${lagMinWindow}, max=${lagMaxWindow}, last=${lagLast})`);
      } else {
        console.error(`[block ${currentBlock}] mute=${(muteRatio * 100).toFixed(1)}% (gain=${gain.toFixed(3)} ${gainMeter}, lag=--)`);
      }
    }

    blockCounter += 1;

    for (let i = 0; i < kBlock; i++) refQ.push(out[i]);
    for (let i = 0; i < kBlock; i++) jitterQ.push(out[i]);
    if (needJitter && jitterQ.length >= jitterSamplesTarget) needJitter = false;

    let speakerBlock = new Int16Array(kBlock);
    if (!needJitter && jitterQ.length >= kBlock) {
      for (let i = 0; i < kBlock; i++) speakerBlock[i] = jitterQ.shift();
    } else {
      speakerBlock.fill(0);
    }
    speakerBlock = applyLoopbackDelayBlock(speakerBlock);
    PortAudio.pushSamplesForPlay(speakerBlock);
  }
}

const tickMs = Math.max(1, Math.round((kBlock / kSr) * 1000));
const timer = setInterval(() => {
  const recBuf = PortAudio.getRecordedSamples();
  const recArr = toInt16Array(recBuf);
  if (recArr.length > 0) {
    enqueueCaptureSamples(recArr);
    if (typeof PortAudio.discardRecordedSamples === 'function') {
      PortAudio.discardRecordedSamples(recArr.length);
    }
  }
  processBlocks();
}, tickMs);

function shutdown() {
  clearInterval(timer);
  try { PortAudio && PortAudio.stopMic && PortAudio.stopMic(); } catch (_) {}
  try { PortAudio && PortAudio.stopSpeaker && PortAudio.stopSpeaker(); } catch (_) {}
}

process.on('SIGINT', () => {
  console.error('stopped.');
  shutdown();
  process.exit(0);
});
